### ISSUE TITLE
worker exit and kernel panic related to UnhandledPromiseRejection in new xs/xsnap

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#6257]
When @mhofman did more-extensive load testing of the `pso-rc0` candidate, he observed a chain halt after about 10 hours of work. We traced it down to the fee-distribution vat worker process exiting (possibly with `rc=0`, but we're not confident of that) in the middle of a delivery. This causes the kernel to panic, because workers spontaneously exiting could be due to memory exhaustion or the host OS shutting down (and killing the worker process before killing the kernel process), and we don't want to e.g. record an in-consensus "vat terminated" consequence for something that might be local to the one validator.

```
stage-26: chain: 2022-09-17T15:00:20.731Z block-manager: block 6714 begin
stage-26: monitor-slog: begin-block 6714
stage-26: chain: 2022-09-17T15:00:20.735Z launch-chain: polled; blockTime:1663426815, h:6714; ADDED = false
stage-26: chain: 2022-09-17T15:00:20.742Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#26)
stage-26: chain: 2022-09-17T15:00:20.742Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#25)
stage-26: chain: 2022-09-17T15:00:20.746Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#27)
stage-26: chain: 2022-09-17T15:00:20.772Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#26)
stage-26: chain: 2022-09-17T15:00:20.773Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#25)
stage-26: chain: 2022-09-17T15:00:20.773Z SwingSet: xsnap: v18: UnhandledPromiseRejectionWarning: (TypeError#27)
stage-26: chain: 2022-09-17T15:00:20.782Z SwingSet: kernel: error in kernel.deliver: (Error#1)
stage-26: chain: Error#1: v18:zcf exited
stage-26: chain:
stage-26: chain:   at packages/xsnap/src/xsnap.js:140:11
stage-26: chain:
stage-26: chain: 2022-09-17T15:00:20.784Z SwingSet: kernel: ##### KERNEL PANIC: error during tryProcessMessage: Error: v18:zcf exited #####
stage-26: chain: 2022-09-17T15:00:20.785Z block-manager: END_BLOCK error: (Error#1)
stage-26: chain: portHandler threw (Error#1)
stage-26: chain: 3:00PM ERR CONSENSUS FAILURE!!! err="Error: v18:zcf exited" module=consensus stack="goroutine 137 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2()\n\t/go/pkg/mod/github.com/agoric-labs/tendermint@v0.34.19-alpha.agoric.1/consensus/state.go:727 +0x4c\npanic({0x7fd17adda980, 0xc00451a590})\n\t/usr/local/go/src/runtime/panic.go:1038 +0x215\ngithub.com/Agoric/agoric-sdk/golang/cosmos/x/swingset.EndBlock({{0x7f...mint/consensus.(*State).handleMsg(0xc000f80000, {{0x7fd17b06aa80, 0xc0026a2628}, {0xc000f2e390, 0x0}})\n\t/go/pkg/mod/github.com/agoric-labs/tendermint@v0.34.19-alpha.agoric.1/consensus/state.go:856 +0x45f\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc000f80000, 0x0)\n\t/go/pkg/mod/github.com/agoric-labs/tendermint@v0.34.19-alpha.agoric.1/consensus/state.go:763 +0x419\ncreated by github.com/tendermint/tendermint/consensus.(*State).OnStart\n\t/go/pkg/mod/github.com/agoric-labs/tendermint@v0.34.19-alpha.agoric.1/consensus/state.go:379 +0x13e\n"
stage-26: runner: Stopping loadgen
```

We weren't able to figure out exactly why it was exiting: we added a bunch of printfs to `xsnap` and didn't manage to find the place that was calling `c_exit`. But we did find that commenting out the [recently-added](https://github.com/agoric-labs/xsnap-pub/commit/84aa01ac86596d46f98eeb5930bcdd33c66e0ef6) call to `fxCheckUnhandledRejections()` at the end of `fxRunLoop` made the process stop exiting.

https://github.com/agoric-labs/xsnap-pub/blob/84aa01ac86596d46f98eeb5930bcdd33c66e0ef6/xsnap/sources/xsnapPlatform.c#L367

We know the vat in question has had a bunch of unhandled rejections (#5745) since June, but [our copy of `fxAbort()`](https://github.com/agoric-labs/xsnap-pub/blob/84aa01ac86596d46f98eeb5930bcdd33c66e0ef6/xsnap/sources/xsnapPlatform.c#L105) merely logs such problems, and explicitly does *not* exit the process (note the lack of `the->abortStatus = status; fxExitToHost(the);` at the end of the `XS_UNHANDLED_REJECTION_EXIT` clause):

https://github.com/agoric-labs/xsnap-pub/blob/84aa01ac86596d46f98eeb5930bcdd33c66e0ef6/xsnap/sources/xsnapPlatform.c#L163-L182

@mhofman is sending a bunch of data to Moddable to see if they can understand why the process is exiting in this case.

While investigating this, we rediscovered #5745, in which some userspace code is doing something like `E('Collateral').offer(..)` (sending a message to a string, not a method-bearing object), and `E(undefined).makeAddCollateralInvitation()` (sending a message to `undefined`). To learn more, we modified the `lockdown` bundle (which includes Endo, hence eventual-send) to add some more logging. The chain does not display worker's stdout/stderr, but when we replayed the vat's transcript, the errors look like:

```
delivery 232 (L 237): ["notify",[["p+216",false,{"body":"{\"brand\":{\"@qclass\":\"slot\",\"iface\":\"Alleged: IST brand\",\"index\":0},\"value\":{\"@qclass\":\"bigint\",\"digits\":\"81671918\"}}","slots":["o-55"]}]]]
xsnap UnhandledPromiseRejectionWarning: (TypeError#26)
xsnap UnhandledPromiseRejectionWarning: (TypeError#25)
xsnap UnhandledPromiseRejectionWarning: (TypeError#27)
xsnap TypeError#27: target has no method undefined , has [ '0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '23', '24', '25', '26', '3', '4', '5', '6', '7', '8', '9', 'length' ]
xsnap TypeError: target has no method "undefined", has ["0","1","10","11","12","13","14","15","16","17","18","19","2","20","21","22","23","24","25","26","3","4","5","6","7","8","9","length"]
 at makeError (/usr/src/agoric-sdk/packages/SwingSet/src/supervisors/subprocess-xsnap/lockdown-subprocess-xsnap.js:2722)
 at fail (/usr/src/agoric-sdk/packages/SwingSet/src/supervisors/subprocess-xsnap/lockdown-subprocess-xsnap.js:2849)
 at localApplyMethod (/usr/src/agoric-sdk/packages/SwingSet/src/supervisors/subprocess-xsnap/lockdown-subprocess-xsnap.js:9177)
 at apply ()
 at dispatchToHandler (/usr/src/agoric-sdk/packages/SwingSet/src/supervisors/subprocess-xsnap/lockdown-subprocess-xsnap.js:9394)
 at win (/usr/src/agoric-sdk/packages/SwingSet/src/supervisors/subprocess-xsnap/lockdown-subprocess-xsnap.js:9734)
 at ()
```

Strangely, when we added a an extra clause to the `target has no method ..` assertion message, the process stopped exiting early. We're hoping that this might help Moddable shed some light on the surprising exit.

@mhofman 's loadgen process is exercising both Vault and AMM trades, neither of which are part of the PSO release. Given our focus in the last two months, it's entirely possible that something has broken in those vats, and/or the loadgen code which exercises them is no longer entirely compatible with them. #5745 has been appropriately scheduled out past PSO, and we don't expect them to be changed or fixed right now.

So, it's possible that there's an XS/xsnap bug/misbehavior that is triggered by non-PSO vats, and that might not be triggered by the PSO operations that are part of this release. If so, this bug might not need to be fixed for PSO (although it's weird enough that I'd be worried about letting it ride).

**Comment by warner** [comment]
I think I figured this out. It's from a surprising (to the handlers) interleaving of now-spontaneously generated UnhandledPromiseRejectionWarning log messages, and the usual delivery/syscall/syscall-result/delivery-result protocol dance.

Our kernel/worker protocol was designed around the idea that the worker is entirely idle between deliveries. The worker does a blocking read of the command pipe, waiting for the kernel to give it a delivery. Then, the worker executes the delivery, performing zero or more syscalls in the process. Each syscall sends a message to the kernel and waits (blocking) for a response (the syscall results). When the worker is done, it sends the delivery results to the kernel, and then goes back to its blocking read. We carefully prevent userspace from regaining control while it's supposed to be idle, so it's impossible for syscalls to appear in between deliveries.

From the kernel's point of view, it can write the delivery message, then do a blocking read for something from the worker. It either receives a syscall request (in which case it executes the syscall and writes a response, then goes back to the blocking read), or it receives the delivery results (in which case it commits the crank and goes on to the next run-queue item, ignoring the worker until it is called upon again).

Later, to get better diagnostics, we enhanced the syscall protocol include `console.log` messages. When the kernel receives a payload from the worker, there are now three cases: syscall, console, or delivery-results. The console messages are sent to the slog and (maybe) to stdout, and do not go into the syscall transcript. We continue to expect that the worker is idle between deliveries (no console logs either).

The kernel is not necessarily that great about ignoring things from the worker while between deliveries. I think we have a flag that would throw an error if a syscall is received, but we might be depending upon some statefullness of the slog for that.

In the error we're seeing now, the basic flow is a delivery (that happened to be a `dispatch.notify()` of some promise results), which didn't provoke any syscalls, but *did* provoke three `UnhandledPromiseRejectionWarning`s, all of which somehow got routed to a `console.log` that reports back up to the kernel. The sequence nominally looked like this:
![bug 6257 - Frame 1](https://user-images.githubusercontent.com/27146/190881838-b8a4becc-f7d8-4726-9ed4-b3e46b1e930e.jpg)

The next thing to know is the exact encoding of the messages that cross between kernel and worker. The first is an xsnap command payload. `xsnap` listens on a pipe and reads complete netstrings from the kernel, then [looks at the first character](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L368-L377) of the decoded payload to see what the parent wants it to do. The three main options are:

* `e`: [`eval()` the rest of the payload as a string](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L392). We use this to install the SES/Endo and supervisor bundles, which register a `handleCommand` function with xsnap for later
* `?`: [send the rest of the payload to `handleCommand()`](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L385-L386). We use this to send vat deliveries into liveslots
* `w`: treat the rest of the payload as a filename, and [write a heap snapshot](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L487) to that path

The second is a worker-to-kernel message. The kernel-side state machine doesn't know ahead of time whether the message it gets from the worker is a syscall request, a log request, or the results of a delivery, so all three must share the same format. These use netstrings again, and the first letter of the payload tells the kernel what to do:

* `.`: the eval (`e`) or handleCommand (`?`) was [successful](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L437). The rest of the payload includes [JSON-encoded metering results](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L744-L751). This is also used to response to `w` snapshot-write commands.
* `!`: the eval/handleCommand [failed somehow](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L442)
* `?`: the worker is [issuing a request](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L788-L800) to the kernel, used for syscalls and log messages

The third is the kernel-to-worker message that responds to syscall/log requests. This is again a netstring, however the payload does not need a first-character dispatch, because the worker is "always" (cue dramatic bug-finding music) in the same state when it arrives. The worker should always be in the [blocking read inside `xs_issueCommand`](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L808).

The correct sequence of messages for our error-provoking delivery looks something like this:
![bug 6257 - Frame 2](https://user-images.githubusercontent.com/27146/190882223-b230f4eb-df0d-4b1e-8643-6826a8aa4d5a.jpg)

The kernel sends the delivery request (`?[\"deliver\", ..`) to the idle worker. While processing that delivery, the worker sends a log request (`?[\"sourcedConsole\", ..`) to the kernel, which responds with the ack (`[\"ok\"]`). The worker does three of these log requests before finishing the delivery and sending the results (`.{\"compute\": ..`). Then the worker goes back to waiting for a new request.

## The Interleaving Bug

What *actually* happened is that the worker somehow decided to send a batch of three log requests *after* the next-to-last delivery completed. All three complain about `UnhandledPromiseRejectionWarning`, probably provoked by the previous delivery. This threw off the two state machines:

![bug 6257 - Frame 3](https://user-images.githubusercontent.com/27146/190882776-7521dd09-3649-4c54-a9a6-5e492e6080c9.jpg)

The kernel's delivery request crossed the worker's out-of-order log request on the wire. From the kernel's point of view, everything started out fine:

* kernel receives results from previous delivery, kernel cycles, finds next delivery to make
* kernel sends delivery request to worker
* kernel receives log request, sends ack
* kernel receives second log request, sends ack
* kernel receives third log request, sends ack
* but then the kernel observes the worker spontaneously exit, with an exit code of `0`

Now from the worker's point of view:

* worker finishes the previous delivery, sends the results
* somehow, after `fxWriteOkay` but before the command parser returns to the top of the loop (maybe in `xsEndHost`?), the engine looks for unhandled promises and finds a few to complain about
* worker sends log request 1, waits for ack
  * worker receives kernel's delivery request (`?[\"deliver\"]`), mis-parses it as a log request ack
  * I think this would cause the `console.log` to fail, because JSON-parsing the leading `?` would fail, but I expect we discard errors during console.log because really, what else could we do
* worker sends log request 2, waits for ack
  * worker receives kernel's ack from log request 1, ok
* worker sends log request 3, waits for ack
  * worker receives kernel's ack from log request 2, ok
* worker finally finishes the command parser loop, returns to a blocking read for a command
  * worker receives kernel's ack from log request 3 (`[\"ok\"]`), mis-parses it as a command
  * the first character `[` is not `e` or `?` or `w`, and this causes xsnap to [exit the command loop](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L519-L521)
  * worker process exits with code `0`


## Fixes

The core problem is that the worker can (now) send log events spontaneously, or at least slightly outside the expected window. The secondary problem is that there are (necessarily) two worker-side parsers, and it's possible to be talking to the wrong one.

The most important fix is to figure out where these logged `UnhandledPromiseRejectionWarnings` are being emitted, and make sure they can't happen after our command processor has sent back its results. I suspect this is happening in the [`fxEndHost()` that we do](https://github.com/agoric-labs/xsnap-pub/blob/916ded7ebd1aa0374cc0dcadf7510b920eb22c08/xsnap/sources/xsnap-worker.c#L440) within the command loop, but I'm not sure.

In the longer run, we might consider:
* Move all log messages to a different pipe, so they aren't interleaved with commands and syscalls
* Encode worker-to-kernel request responses in the same language as the kernel-to-worker commands, but with a distinctive prefix character (not `e`/`?`/`w`/etc), and bail loudly if one parser gets a message meant for the other. This would have sped up the diagnostic process considerably
* Allow userspace and liveslots to send log messages via the pipes, but not XS internal code. I am strongly suspicious of the following clause in `supervisor-subprocess-xsnap.js` as a possible source of these warnings:

https://github.com/Agoric/agoric-sdk/blob/2a66893703d75d76b58aee21722268d0b9bcfd86/packages/SwingSet/src/supervisors/subprocess-xsnap/supervisor-subprocess-xsnap.js#L319-L322

**Comment by warner** [comment]
I've been able to confirm that the messages are happening during that `fxEndHost()`, not the `fxRunLoop()`. I don't understand why commenting out the `fxCheckUnhandledRejections()` in `fxRunLoop()` makes them stop happening, but it does.

**Comment by mhofman** [comment]
Fixed by #6259
